File: python/llm_sketchkit/canon.py
# Code authors: Vijay Erramilli and Codex
"""Canonicalization profiles for llm-sketchkit."""

from __future__ import annotations

import unicodedata

TEXT_V1 = "text_v1"


class CanonicalizationError(ValueError):
    """Base class for canonicalization failures."""


class UnsupportedProfileError(CanonicalizationError):
    """Raised when a canonicalization profile is not implemented."""


class InvalidUTF8Error(CanonicalizationError):
    """Raised when input bytes are not valid UTF-8."""


def canonicalize(profile: str, value: bytes | str) -> bytes:
    """Canonicalize a UTF-8 text value under the named profile."""

    if profile != TEXT_V1:
        raise UnsupportedProfileError(profile)
    text = _decode_text(value)
    return _canonicalize_text_v1(text).encode("utf-8")


def canonicalize_text_v1(value: bytes | str) -> bytes:
    """Canonicalize a UTF-8 text value under `text_v1`."""

    return canonicalize(TEXT_V1, value)


def _decode_text(value: bytes | str) -> str:
    if isinstance(value, str):
        try:
            value.encode("utf-8")
        except UnicodeEncodeError as exc:
            raise InvalidUTF8Error("invalid UTF-8 string") from exc
        return value

    try:
        return value.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise InvalidUTF8Error("invalid UTF-8 bytes") from exc


def _canonicalize_text_v1(value: str) -> str:
    normalized = unicodedata.normalize("NFC", value)
    normalized = normalized.replace("\r\n", "\n").replace("\r", "\n")
    normalized = normalized.strip()
    return unicodedata.normalize("NFC", normalized)

File: go/sketchkit/canon/canon.go
// SPDX-License-Identifier: Apache-2.0
// Code authors: Vijay Erramilli and Codex
// Package canon implements llm-sketchkit canonicalization profiles.
package canon

import (
	"errors"
	"fmt"
	"strings"
	"unicode"
	"unicode/utf8"

	"golang.org/x/text/unicode/norm"
)

// Profile names a canonicalization profile from spec/canonicalization.md.
type Profile string

const (
	// TextV1 applies NFC normalization, newline normalization, and Unicode trim.
	TextV1 Profile = "text_v1"
)

var (
	// ErrUnsupportedProfile reports a profile that is not implemented.
	ErrUnsupportedProfile = errors.New("unsupported canonicalization profile")

	// ErrInvalidUTF8 reports input that cannot be decoded as UTF-8.
	ErrInvalidUTF8 = errors.New("invalid utf-8 input")
)

// Canonicalize applies profile to UTF-8 bytes and returns canonical UTF-8 bytes.
func Canonicalize(profile Profile, input []byte) ([]byte, error) {
	if profile != TextV1 {
		return nil, fmt.Errorf("%w: %s", ErrUnsupportedProfile, profile)
	}
	if !utf8.Valid(input) {
		return nil, ErrInvalidUTF8
	}

	return canonicalizeTextV1(string(input)), nil
}

// CanonicalizeString applies profile to a Go string and returns canonical UTF-8 bytes.
func CanonicalizeString(profile Profile, input string) ([]byte, error) {
	if profile != TextV1 {
		return nil, fmt.Errorf("%w: %s", ErrUnsupportedProfile, profile)
	}
	if !utf8.ValidString(input) {
		return nil, ErrInvalidUTF8
	}

	return canonicalizeTextV1(input), nil
}

func canonicalizeTextV1(input string) []byte {
	normalized := norm.NFC.String(input)
	normalized = strings.ReplaceAll(normalized, "\r\n", "\n")
	normalized = strings.ReplaceAll(normalized, "\r", "\n")
	normalized = strings.TrimFunc(normalized, unicode.IsSpace)
	normalized = norm.NFC.String(normalized)

	return []byte(normalized)
}
